language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -276,136 +357,83 @@ computeCek con t@(Var _ varName)   = do
-    let cost = memoryUsage ()
-    argName <- freshName "arg"
-    withVarEnv (extendVarEnv argName arg argVarEnv funVarEnv) $ k (Var cost argName)
 
 -- | Apply a function to an argument and proceed.
 -- If the function is a 'LamAbs', then extend the current environment with a new variable and proceed.
--- If the function is not a 'LamAbs', then 'Apply' it to the argument and view this
--- as an iterated application of a 'BuiltinName' to a list of 'Value's.
--- If succesful, proceed with either this same term or with the result of the computation
--- depending on whether 'BuiltinName' is saturated or not.
+-- If the function is a 'Builtin', then check whether we've got the right number of arguments.
+-- If we do, then ask the constant application machinery to apply it, and proceed with
+-- the result (or throw an error if something goes wrong); if we don't, then add the new
+-- argument to the VBuiltin and call returnCek to look for more arguments.
 applyEvaluate
     :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => VarEnv uni
-    -> VarEnv uni
-    -> Context uni
-    -> WithMemory Value uni
-    -> WithMemory Value uni
+    => Context uni
+    -> CekVal uni   -- lsh of application
+    -> CekVal uni   -- rhs of application
     -> CekM uni (Plain Term uni)
-applyEvaluate funVarEnv argVarEnv con (LamAbs _ name _ body) arg =
-    withVarEnv (extendVarEnv name arg argVarEnv funVarEnv) $ computeCek con body
-applyEvaluate funVarEnv argVarEnv con fun arg = do
-    withScopedArgIn funVarEnv argVarEnv arg $ \arg' ->
-        let term = Apply (memoryUsage () <> memoryUsage fun <> memoryUsage arg') fun arg'
-        in case termAsPrimIterApp term of
-            Nothing                       ->
-                throwingWithCause _MachineError NonPrimitiveApplicationMachineError $ Just term
-            Just (IterApp headName spine) -> do
-                constAppResult <- applyStagedBuiltinName headName spine
-                case constAppResult of
-                    ConstAppFailure     ->
-                        throwingWithCause
-                            _EvaluationError
-                            (UserEvaluationError CekEvaluationFailure)
-                            (Just term)
-                    ConstAppSuccess res -> computeCek con res
-                    ConstAppStuck       -> returnCek con term
+applyEvaluate ctx (VLamAbs _ name _ty body env) arg =
+    withEnv (extendEnv name arg env) $ computeCek ctx body
+applyEvaluate ctx val@(VBuiltin ex argEnv bn count tyargs args) arg = withEnv argEnv $ do
+    let args' = arg:args
+        count' = count - 1
+    if count' /= 0
+        then returnCek ctx $ VBuiltin ex argEnv bn count' tyargs args'
+        else do
+            res <- applyStagedBuiltinName bn (reverse args')
+            case res of
+                EvaluationSuccess t -> returnCek ctx t  -- NOTE that this is 'returnCek' now.
```

Comment:
I wrote that note just to draw attention to the change, I don't think we should keep it in the code.

Reply:
OK, I took that out. We do keep switching back and forth between builtins returning terms and values though!

Reply:
> OK, I took that out. We do keep switching back and forth between builtins returning terms and values though!

What's the latest on returning values vs returning terms? At the risk of applying my prejudices in the wrong place, I think it's inefficient to return terms that the CEK machine has to reprocessed to rediscover that there were values all along. Indeed, I have seen the CEK machine motivated by saying that one of the reasons to delay substitution is that substitution returns a term which needs to be reprocessed.

Reply:
@jmchapman at the moment we can't really support returning arbitrary terms in your version of the CEK machine, so we're pretty much bound to returning values. I've commented on that recently somewhere, but with the amount of discussion going on I now can't find anything.